JS.md

```diff
@@ -224,24 +270,204 @@ WebAssembly Exported Functions have a `[[Call]](this, argValues)` method defined
    passing [[Instance]], [[FunctionIndex]], and `args`.
- * Coerce the result of `Eval.invoke` as follows:
-   * if the return value is `None`, return `undefined`
-   * interpret `i32` as a signed integer and convert that integer to a Number value
-   * throw a `TypeError` if returning an `i64`
-   * return `f32`/`f64` as Number values, possibly performing
-     [canonicalization of NaNs](http://tc39.github.io/ecma262/#sec-setvalueinbuffer)
+ * Return [`ToJSValue`](#tojsvalue)`(ret)`.
 
 Exported Functions do not have a [[Construct]] method and thus it is not possible to 
 call one with the `new` operator.
 
+## `WebAssembly.Memory` Objects
+
+A `WebAssembly.Memory` object contains a single [linear memory](AstSemantics.md#linear-memory)
+which can be simultaneously referenced by multiple `Instance` objects. Each
+`Memory` object has two internal slots:
+ * [[Memory]] : a [`Memory.memory`](https://github.com/WebAssembly/spec/blob/master/ml-proto/spec/memory.mli)
+ * [[BufferObject]] : the current `ArrayBuffer` whose [[ArrayBufferByteLength]]
+   matches the current byte length of [[Memory]]
+
+### `WebAssembly.Memory` Constructor
+
+The `WebAssembly.Memory` constructor has the signature:
+```
+new Memory(memoryDescriptor)
+```
+If the NewTarget is `undefined`, a `TypeError` exception is thrown (i.e., this
+constructor cannot be called as a function without `new`).
+
+If `Type(memoryDescriptor)` is not Object, a `TypeError` is thrown.
+
+Let `initial` be [`ToInteger`](http://tc39.github.io/ecma262/#sec-tointeger)([`Get`](http://tc39.github.io/ecma262/#sec-get-o-p)(`memoryDescriptor`, `"initial"`)).
+
+If [`HasProperty`](http://tc39.github.io/ecma262/#sec-hasproperty)(`"maximum"`),
+then let `maximum` be [`ToInteger`](http://tc39.github.io/ecma262/#sec-tointeger)([`Get`](http://tc39.github.io/ecma262/#sec-get-o-p)(`memoryDescriptor`, `"maximum"`)).
+Otherwise, let `maximum` be `None`.
+
+Let `memory` be the result of calling 
+[`Memory.create`](https://github.com/WebAssembly/spec/blob/master/ml-proto/spec/memory.mli#L18)
+given arguments `initial` and `maximum`. Note that `initial` and `maximum` are
+specified in units of WebAssembly pages (64KiB).
+
+(Note: the ML spec currently doesn't implement the maximum memory limit; we
+assume here it will be extended in the future.)
+
+Return the result of [`CreateMemoryObject`](#creatememoryobject)(`memory`).
+
+### CreateMemoryObject
+
+Given a [`Memory.memory`](https://github.com/WebAssembly/spec/blob/master/ml-proto/spec/memory.mli)
+`m`, to create a `WebAssembly.Memory`:
+
+Let `buffer` be a new `ArrayBuffer` whose
+[[[ArrayBufferData]]](http://tc39.github.io/ecma262/#sec-properties-of-the-arraybuffer-prototype-object)
+aliases `m` and whose 
+[[[ArrayBufferByteLength]]](http://tc39.github.io/ecma262/#sec-properties-of-the-arraybuffer-prototype-object)
+is set to the byte length of `m`.
+
+Return a new `WebAssembly.Memory` instance with [[Memory]] set to `m` and
+[[BufferObject]] set to `buffer`.
+
```

Comment:
Yes: the goal is to have 100% symmetry between a `Memory` constructor and the linear memory section in modules: how a memory is created should be orthogonal from how it is used.